src/acl_kernel_if.cpp

```diff
@@ -400,20 +400,42 @@ static uintptr_t acl_kernel_cra_set_segment_rom(acl_kernel_if *kern,
 
 static int acl_kernel_cra_read(acl_kernel_if *kern, unsigned int accel_id,
                                unsigned int addr, unsigned int *val) {
+  int result = 0;
   assert(kern->cra_ring_root_exist);
-  uintptr_t segment_offset = acl_kernel_cra_set_segment(kern, accel_id, addr);
-  acl_assert_locked_or_sig();
-  return acl_kernel_if_read_32b(
-      kern, (unsigned)OFFSET_KERNEL_CRA + (unsigned)segment_offset, val);
+#ifdef __linux__
+  acl_sig_block_signals();
+#endif
+  {
+    std::lock_guard<std::mutex> lock(kern->segment_mutex);
+    uintptr_t segment_offset = acl_kernel_cra_set_segment(kern, accel_id, addr);
+    acl_assert_locked_or_sig();
+    result = acl_kernel_if_read_32b(
+        kern, (unsigned)OFFSET_KERNEL_CRA + (unsigned)segment_offset, val);
+  }
+#ifdef __linux__
+  acl_sig_unblock_signals();
+#endif
+  return result;
 }
 
 int acl_kernel_cra_read_64b(acl_kernel_if *kern, unsigned int accel_id,
                             unsigned int addr, uint64_t *val) {
+  int result = 0;
   assert(kern->cra_ring_root_exist);
-  uintptr_t segment_offset = acl_kernel_cra_set_segment(kern, accel_id, addr);
-  acl_assert_locked_or_sig();
-  return acl_kernel_if_read_64b(
-      kern, (unsigned)OFFSET_KERNEL_CRA + (unsigned)segment_offset, val);
+#ifdef __linux__
+  acl_sig_block_signals();
+#endif
+  {
```

Comment:
Is there a reason why you need to put this into a new scope?  Does the mutex need to be release before the unlock?  If you do as I suggested, that would be done as well

Reply:
The new scope is for the mutex to unlock when exiting the scope, but yea I'll try the change you suggested above, maybe that way I can avoid this new scope.

Reply:
You can remove the scope.  C++ rules say that the mutex will be released before the unblock of the signals

Reply:
Ah is it because of this?
https://stackoverflow.com/questions/2254263/order-of-member-constructor-and-destructor-calls

Reply:
yes.  they will be destructed in reverse order from the construction

Reply:
Scope removed